test/object-store/c_api/c_api.cpp

```diff
@@ -5800,4 +5801,108 @@ TEST_CASE("app: flx-sync basic tests", "[c_api][flx][sync]") {
+    public:
+        TestSyncTimer(DefaultSocketProvider& socket_provider, std::chrono::milliseconds delay,
+                      SyncSocketProvider::FunctionHandler&& handler)
+        {
+            m_timer = socket_provider.create_timer(delay, std::move(handler));
+        }
+
+        void cancel() override
+        {
+            m_timer->cancel();
+        }
+
+    private:
+        SyncSocketProvider::SyncTimer m_timer;
+    };
+
+    auto logger = std::make_shared<util::StderrLogger>();
+    DefaultSocketProvider default_socket_provider(logger, "SocketProvider");
+
+    auto post_fn = [](realm_userdata_t userdata, realm_sync_socket_callback_t* callback) {
+        auto socket_provider = static_cast<DefaultSocketProvider*>(userdata);
+        auto cb = [callback_copy = callback](Status s) {
+            realm_sync_socket_callback_complete(callback_copy, static_cast<status_error_code_e>(s.code()),
+                                                s.reason().c_str());
+        };
+        socket_provider->post(std::move(cb));
+    };
+    auto create_timer_fn = [](realm_userdata_t userdata, uint64_t delay_ms,
+                              realm_sync_socket_callback_t* callback) -> realm_sync_socket_timer_t {
+        auto socket_provider = static_cast<DefaultSocketProvider*>(userdata);
+        return static_cast<realm_sync_socket_timer_t>(
+            new TestSyncTimer(*socket_provider, std::chrono::milliseconds(delay_ms), std::move(**callback)));
+    };
+    auto cancel_timer_fn = [](realm_userdata_t, realm_sync_socket_timer_t timer_userdata) {
+        auto timer = static_cast<TestSyncTimer*>(timer_userdata);
+        timer->cancel();
+    };
+    auto free_timer_fn = [](realm_userdata_t, realm_sync_socket_timer_t timer_userdata) {
+        realm_release(timer_userdata);
+    };
+    auto websocket_connect_fn =
+        [](realm_userdata_t userdata, realm_websocket_endpoint_t endpoint,
+           realm_websocket_observer_t* realm_websocket_observer) -> realm_sync_socket_websocket_t {
+        auto socket_provider = static_cast<DefaultSocketProvider*>(userdata);
+        return static_cast<realm_sync_socket_websocket_t>(
+            new TestWebSocket(*socket_provider, endpoint, realm_websocket_observer));
+    };
+    auto websocket_async_write_fn = [](realm_userdata_t, realm_sync_socket_websocket_t websocket_userdata,
+                                       const char* data, size_t size, realm_sync_socket_callback_t* callback) {
+        auto websocket = static_cast<TestWebSocket*>(websocket_userdata);
+        websocket->async_write_binary(util::Span{data, size}, std::move(**callback));
+    };
+    auto websocket_free_fn = [](realm_userdata_t, realm_sync_socket_websocket_t websocket_userdata) {
+        realm_release(websocket_userdata);
+    };
+
+    auto socket_provider = realm_sync_socket_new(
+        static_cast<realm_userdata_t>(&default_socket_provider), [](realm_userdata_t) {}, post_fn, create_timer_fn,
```

Comment:
For the userdata free function, you should send in an int in the captures and then increment it in the lambda body so you can verify that free is called exactly once when `realm_release(socket_provider)` is called.
Maybe also check to verify that it is being called with the correct value.
example:
```
int free_count = 0;
. . .
[&](realm_userdata_t user_ptr) {
    free_count++;
    REQUIRE(user_ptr == &default_socket_provider);
}
. . .
realm_release(socket_provider);
REQUIRE(free_count == 1);
```

Reply:
Cannot do that unfortunately since a capturing lambda cannot be cast to a function pointer. I can work around that though by wrapping default_socket_provider and any test data into its own type.

Reply:
That's right - I was trying to remember why I created the special structure for each of my tests...
You could create a local struct that contains the socket_provider and a free_count and use that as your user data... I have something like that modeled in PR #6156 that maybe you could take advantage of...
The main thing is to confirm that the userdata free works as expected so there's no leaks or double frees.